site/guide.md

```diff
@@ -437,17 +437,146 @@ import cats.effect.unsafe.implicits.global
+
+```scala 
+val finite = Stream('a', 'b', 'c', 'd', 'e').covary[IO]
+val infinite = Stream.iterate(0)(_ + 1).covary[IO]
+
+// Left --------- Right
+finite.mergeHaltL(infinite)     // Terminates
+infinite.mergeHaltL(finite)     // Doesn't terminate
+
+finite.mergeHaltR(infinite)     // Doesn't terminate
+infinite.mergeHaltR(finite)     // Terminates
+
+finite.mergeHaltBoth(infinite)  // Terminates
+infinite.mergeHaltBoth(finite)  // Also terminates
+```
+
+Since it is quite common to terminate one stream as soon as the other is finished (as in a producer-consumer environment), there are optimisations over the `merge` variants. `concurrently` is one of them, as it will terminate the resulting stream when the stream on the left halts. The stream on the right will also terminate at that point, discarding its values in the meantime (similar to `finite.mergeHaltL(infinite.drain)`):
+
+```scala mdoc
+import cats.effect.IO
+import cats.effect.std.{Queue, Random}
+import cats.effect.unsafe.implicits.global
+import scala.concurrent.duration._
+
+def producer(queue: Queue[IO, Option[Int]])(implicit rnd: Random[IO]): Stream[IO, Option[Int]] = 
+  Stream
+    .repeatEval(Random[IO].betweenInt(100,800))
+    .evalTap(n => IO.println(s"Produced: $n"))
+    .flatMap(t => Stream.sleep[IO](t.milliseconds) >> Stream.emit(if (t >= 750) None else Some(t)))
+    .evalTap(queue.offer)
+
+def consumer(queue: Queue[IO, Option[Int]]): Stream[IO, Unit] = 
+  Stream.fromQueueNoneTerminated(queue, 10).evalMap(n => IO.println(s"Consumed: $n"))
+
+val concurrentlyDemo = 
+  Stream.eval(Queue.bounded[IO, Option[Int]](20)).flatMap { queue =>
+    Stream.eval(Random.scalaUtilRandom[IO]).flatMap { implicit rnd =>
+
+      consumer(queue).concurrently(producer(queue))
+
+    }
+  }
+
+concurrentlyDemo.compile.drain.unsafeRunSync()
+```
+
+In the example above, the `consumer` stream will terminate if an element produced by the `producer` takes more than 750 milliseconds.
+
+
+The `parEvalMap` function allows you to evaluate effects in parallel and emit the results in order on up to `maxConcurrent` fibers at the same time, similar to the `parTraverseN` method that you would normally use in standard library collections:
+
+``` scala mdoc
+import cats.effect.IO
+import cats.effect.unsafe.implicits.global
+
+// This will evaluate 5 effects in parallel
+Stream(1, 2, 3, 4, 5).parEvalMap(5)(n => IO.pure(n * 2)).compile.toVector.unsafeRunSync()
+```
+
+However, its use with pure operations is rare; it is more common with functions or combinators that can have side effects:
+
+```scala mdoc
+import cats.effect.IO
+import cats.effect.unsafe.implicits.global
+import fs2.io.file.{Path, Files}
+
+val paths = List(
+  Path("sample_file_part1.txt"),
+  Path("sample_file_part2.txt"),
+  Path("sample_file_part3.txt"),
+).map(Path("testdata") / _)
+
+def loadFile(path: Path): IO[String] = 
```

Comment:
I don't love this example for a couple of reasons:
- concurrency not really adding any benefit given all actions are against the file system
- unnecessarily parses the file contents to a string
- unnecessarily loads full file contents in memory
- if those downsides aren't important, then the same thing could be done entirely with `IO` and without `Stream`

Here's a full streaming version of this file concat routine that maintains a streaming nature and avoids parsing:

```scala
def concat(inPaths: Stream[IO, Path], outPath: Path): Stream[IO, Nothing] =
  inPaths.flatMap(p => Files[IO].readAll(p)).through(Files[IO].writeAll(outPath))
```

Reply:
You are right, I did It just as an excuse to maintain the order of the original stream 🤔 let me thing of a better example!